File: src/Classes/Item.ts
import { ILeveled } from "../Interfaces/ILeveled";
import { IPrototype } from "../Interfaces/iPrototype";

import { EType } from "../Enums/EType";
import { EWeapon } from "../Enums/EWeapon";

export class Item implements ILeveled, IPrototype {

    name:string
    type:EType 
    level:number 
    scope:number
    duration:number 
    damage :number
    exploRange :number
    weaponType :EWeapon
    ammo :number
    textureMap: Map<number, string>
    currentTexture: string

    constructor(
        name:string,
        type:EType ,
        level:number, 
        scope:number,
        duration:number, 
        damage :number,
        exploRange :number,
        weaponType :EWeapon,
        ammo :number,
        textureMap: Map<number, string>, 
        currentTexture: string,
        ){
        this.name = name;
        this.type = type;
        this.level = level;
        this.scope = scope;
        this.duration = duration;
        this.damage = damage;
        this.exploRange = exploRange;
        this.weaponType = weaponType;
        this.ammo = ammo;
        this.textureMap = textureMap;
        this.currentTexture = currentTexture;
    }

    getName():string{
        return this.name;
    }
    getType():EType{
        return this.type;
    }
    getLevel():number{
        return this.level;
    }
    getScope():number{
        return this.scope;
    }
    getDamage():number{
        return this.damage;
    }
    getTextureMap(): Map<number, string>{
        return this.textureMap;
    }
    //
   
    
    getDuration():number{
        return this.duration;
    }
    levelUp(): void {
        this.level += 1; //set levelUp
        
    }
    levelDown(): void {
        this.level -=1;
    }
 

    toJson(): string {

        const textureMapObject: any = {};
        this.textureMap.forEach((value, key) => {
            textureMapObject[key] = value;
        });

        let objectJson: any =  {
            name: this.name,
            type: this.type,
            level: this.level,
            scope: this.scope,
            duration: this.duration,
            damage: this.damage,
            exploRange: this.exploRange,
            weaponType: this.weaponType,
            ammo: this.ammo,
            textureMap: textureMapObject,
            currentTexture: this.currentTexture
        }

        return JSON.stringify(objectJson);

    }


    // Declarations in order to be able to use the methods in Character class before they are implemented
    clone(): Item{
        return new Item(this.name, 
            this.type, this.level,
            this.scope,
            this.duration, 
            this.damage, 
            this.exploRange, 
            this.weaponType, 
            this.ammo, 
            this.textureMap, 
            this.currentTexture);
    }
    deepClone(): Item{
        const textureMap = new Map<number, string>();
        this.textureMap.forEach((value, key) => {
            textureMap.set(key, value);
        });

        return  new Item(this.name, 
            this.type, this.level,
            this.scope,
            this.duration, 
            this.damage, 
            this.exploRange, 
            this.weaponType, 
            this.ammo, 
            textureMap, 
            this.currentTexture);
    }

  }
  

File: Game/CharacterSimulator/src/Model/CharactersLibrary/Classes/Item.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package Model.CharactersLibrary.Classes;

import Model.CharactersLibrary.Intefaces.ILeveled;
import Model.Enums.EType;
import Model.Enums.EWeapon;

import java.util.HashMap;

/**
 *
 * @author Esteb
 */
public class Item implements ILeveled{
    public String name;
    public EType type;
    public double level;
    public double scope;
    public double duration;
    public double damage;
    public double exploRange;
    public EWeapon weaponType;
    public double ammo;
    public HashMap<Integer,String> textureMap;
    public String currentTexture;

    public Item(String name, EType type, double level, double scope, double duration, double damage, double exploRange, EWeapon weaponType, double ammo, HashMap<Integer, String> textureMap, String currentTexture) {
        this.name = name;
        this.type = type;
        this.level = level;
        this.scope = scope;
        this.duration = duration;
        this.damage = damage;
        this.exploRange = exploRange;
        this.weaponType = weaponType;
        this.ammo = ammo;
        this.textureMap = textureMap;
        this.currentTexture = currentTexture;
    }


    @Override
    public void levelUp() {
        this.level+=1;
    }

    @Override
    public void levelDown() {
        this.level-=1;
    }

    public String getName() {
        return name;
    }

    public int getLevel() {
        return (int)level;
    }

    public int getScope() {
        return (int) scope;
    }

    public double getDuration() {
        return duration;
    }

    public double getDamage() {
        return damage;
    }

    public int getExploRange() {
        return (int)exploRange;
    }

    public int getAmmo() {
        return (int) ammo;
    }

    public String getCurrentTexture() {
        return currentTexture;
    }

    @Override
    public String toString() {
        return "Item{" +
                "name='" + name + '\'' +
                ", type=" + type +
                ", level=" + level +
                ", scope=" + scope +
                ", duration=" + duration +
                ", damage=" + damage +
                ", exploRange=" + exploRange +
                ", weaponType=" + weaponType +
                ", ammo=" + ammo +
                ", textureMap=" + textureMap +
                ", currentTexture='" + currentTexture + '\'' +
                '}';
    }

}
